In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [4]:
customer = pd.read_json('/Users/erica/Desktop/itmgt/GITHUB/dy-itm25/dim_customer.json')
invoice = pd.read_csv('/Users/erica/Desktop/itmgt/GITHUB/dy-itm25/fct_invoice.csv')

In [5]:
customer

age first_name gender     id last_name
0     48.0   Nicholas      M      0    Flores
1      NaN    Jeffery      M      1      Rowe
2     57.0     Alexis      F      2    Benton
3     73.0        Amy      F      3   Johnson
4     20.0      David      M      4     Moore
...    ...        ...    ...    ...       ...
5186   NaN     Arthur      M   9428    Harris
5187   NaN      David      M   9494       Roy
5188  85.0     Thomas      M  10350    Turner
5189   NaN      April      F  10587   Sanchez
5190   NaN     Andrew      M  10934   Compton

[5191 rows x 5 columns]

# Easy
<font color=7EBC80>**How many unique customers are in the dataset?**</font>

In [6]:
customer_id = invoice['customer_id']

customer_id.nunique()

5191

<u>*answer: there are 5191 unique customers in the dataset*<u>

<font color=7EBC80>**What are the different categories of products available? How many unique categories are there?**</font>

In [12]:
category = invoice['category']

category.unique()

array(['Clothing', 'Shoes', 'Books', 'Cosmetics', 'Food & Beverage',
       'Toys', 'Technology', 'Souvenir'], dtype=object)

In [10]:
category.nunique()

8

<u>*answer: there are 8 unique categories - clothing, shoes, books, cosmetics, food & bevarage, toys, technology, and souvenir*<u>

<font color=7EBC80>**Which payment method is the most popular? How many times was it used?**</font>

In [15]:
payment_method = invoice['payment_method']

In [18]:
payment_method.value_counts()

payment_method
Cash           44447
Credit Card    34931
Debit Card     20079
Name: count, dtype: int64

<u>*answer: cash is the most popular payment method, used 44,447 times*<u>

# Medium
<font color=7EBC80>**What are the three most popular categories, by total sales?**</font>

In [9]:
price = invoice['price']
quantity = invoice['quantity']

total_price = price * quantity

invoice.insert(5,'total_price',total_price,True)

In [47]:
invoice.groupby('category').sum()['total_price'].rank(ascending=False)

category
Books              7.0
Clothing           1.0
Cosmetics          4.0
Food & Beverage    6.0
Shoes              2.0
Souvenir           8.0
Technology         3.0
Toys               5.0
Name: total_price, dtype: float64

<u>*answer: the three most popular categories by total sales are clothing, shoes, and technology*<u>

<font color=7EBC80>**What are the total sales attributed to customers over the age of 45?**</font>

In [10]:
merged = pd.merge(
    invoice , 
    customer,
    how = 'inner',
    left_on = 'customer_id',
    right_on = 'id'
)

In [55]:
age = merged['age']

over_45 = age > 45

merged[over_45].sum()['total_price']

82039768.15

<u>*answer: the total sales attributed to customers over the age of 45 is 82,039,768.15*<u>

<font color=7EBC80>**How is the data distributed across different invoice dates? Are there any seasonal trends or patterns? (Use a graph for this.)**</font>

In [99]:
merged.sort_values(by='invoice_date', ascending=True)

invoice_no  customer_id         category  quantity    price  \
40380    I113208          233       Technology         2  2100.00   
5432     I474975           89         Clothing         2   600.16   
92987    I371949         1818             Toys         2    71.68   
62130    I260437          968        Cosmetics         2    81.32   
48139    I301161         1071        Cosmetics         3   121.98   
...          ...          ...              ...       ...      ...   
89658    I323106         1275        Cosmetics         2    81.32   
51515    I307740          567        Cosmetics         3   121.98   
36542    I268432          379  Food & Beverage         3    15.69   
61732    I259307         1938             Toys         1    35.84   
24842    I513689          986  Food & Beverage         3    15.69   

       total_price payment_method invoice_date     shopping_mall   age  \
40380      4200.00           Cash   2021-01-01    Forum Istanbul   NaN   
5432       1200.32    Credit Card   2021-01-01      Metropol AVM   NaN   
92987       143.36    Credit Card   2021-01-01         Metrocity  81.0   
62130       162.64           Cash   2021-01-01  Mall of Istanbul  43.0   
48139       365.94     Debit Card   2021-01-01  Mall of Istanbul  45.0   
...            ...            ...          ...               ...   ...   
89658       162.64           Cash   2023-03-08            Kanyon   NaN   
51515       365.94     Debit Card   2023-03-08    Viaport Outlet  20.0   
36542        47.07           Cash   2023-03-08            Kanyon   NaN   
61732        35.84           Cash   2023-03-08      Metropol AVM   NaN   
24842        47.07           Cash   2023-03-08            Kanyon  60.0   

      first_name gender    id    last_name    month  
40380      Brian      M   233      Coleman  January  
5432     Michael      M    89   Mccullough  January  
92987     Joshua      M  1818  Christensen  January  
62130    Michael      M   968        Brown  January  
48139   Jonathan      M  1071        Perez  January  
...          ...    ...   ...          ...      ...  
89658      Kevin      M  1275      Charles    March  
51515      David      M   567         Luna    March  
36542      Jesse      M   379        Mcgee    March  
61732       Eric      M  1938        Moran    March  
24842    Michael      M   986       Wilson    March  

[99457 rows x 15 columns]

In [96]:
merged['invoice_date'] = pd.to_datetime(merged['invoice_date'],format='mixed', dayfirst=True)

date = merged['invoice_date']

In [98]:
merged.insert(14,'month',date.dt.month_name(),True)

In [ ]:
merged[['category','total_price','month']].groupby(['month']).sum()['total_price']

In [ ]:
month = merged['month'].values.tolist()
total_price = merged['total_price'].values.tolist()

plt.plot(month, total_price, 'b-', label='sales')
plt.legend(loc='best')
plt.show()

#did not work, will try again T-T


# Hard
<font color=7EBC80>**Create a pivot table showing the breakdown of sales across these dimensions, in this order: category, decade age range (e.g., 10-19, 20-29, and so forth).**</font>

In [ ]:
merged

In [12]:
categories = list(merged['category'].unique())

labels = []
for i in range(10):
    start_age = i * 10
    end_age = start_age + 9
    range_label = f"{start_age}-{end_age}"
    labels.append(range_label)

    bins = [0, 9, 19, 29, 39, 49, 59, 69, 79, 89, 99]

merged['age_range'] = pd.cut(merged['age'], 
                             bins=bins, 
                             labels=labels, 
                             right=False)

pd.pivot_table(merged, values='total_price', 
               index='category',
               columns='age_range', 
               aggfunc='sum')

age_range        0-9       10-19       20-29        30-39        40-49  \
category                                                                 
Books            0.0     7923.45    76022.70     87157.95     87885.15   
Clothing         0.0  1144505.12  9959955.28  11540776.72  11447451.84   
Cosmetics        0.0    65991.18   583796.28    731310.76    682396.78   
Food & Beverage  0.0     9434.92    68633.29     87911.07     91922.48   
Shoes            0.0   634379.69  5571978.28   7303468.73   7163028.95   
Souvenir         0.0     6181.71    54438.93     63834.66     68444.55   
Technology       0.0   739200.00  5053650.00   6298950.00   6082650.00   
Toys             0.0    35911.68   366392.32    424704.00    408253.44   

age_range             50-59        60-69       70-79       80-89  90-99  
category                                                                 
Books              72735.15     81885.75    57706.35    29224.35    0.0  
Clothing         9943150.80  11083154.72  8061649.20  4812382.96    0.0  
Cosmetics         563954.20    669832.84   443722.58   274495.66    0.0  
Food & Beverage    72132.16     81488.63    58256.97    33879.94    0.0  
Shoes            5778436.76   6131336.72  4675324.30  2827400.87    0.0  
Souvenir           54966.78     57418.35    38216.34    26662.29    0.0  
Technology       4458300.00   5421150.00  3808350.00  2134650.00    0.0  
Toys              330516.48    370693.12   293888.00   162856.96    0.0